File: dailyques/Q85.cpp
//Q85.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int m=matrix.size(), n=matrix[0].size(), ans=0;
        vector<int> h(n,0);
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++) h[j]= (matrix[i][j]=='1')? h[j]+1:0;
            stack<int> st; vector<int> l(n), r(n);
            for(int j=0;j<n;j++){
                while(!st.empty() && h[st.top()]>=h[j]) st.pop();
                l[j]= st.empty()? -1: st.top();
                st.push(j);
            }
            while(!st.empty()) st.pop();
            for(int j=n-1;j>=0;j--){
                while(!st.empty() && h[st.top()]>=h[j]) st.pop();
                r[j]= st.empty()? n: st.top();
                st.push(j);
            }
            for(int j=0;j<n;j++) ans=max(ans,h[j]*(r[j]-l[j]-1));
        }
        return ans;
    }
};

File: dailyques/Q85.py
#Q85.py
class Solution(object):
    def maximalRectangle(self, matrix):
        m,n=len(matrix),len(matrix[0])
        h=[0]*n; ans=0
        for i in range(m):
            for j in range(n):
                h[j]=h[j]+1 if matrix[i][j]=='1' else 0
            st=[]; l=[0]*n; r=[0]*n
            for j in range(n):
                while st and h[st[-1]]>=h[j]: st.pop()
                l[j]=st[-1] if st else -1
                st.append(j)
            st=[]
            for j in range(n-1,-1,-1):
                while st and h[st[-1]]>=h[j]: st.pop()
                r[j]=st[-1] if st else n
                st.append(j)
            for j in range(n):
                ans=max(ans,h[j]*(r[j]-l[j]-1))
        return ans
